pkg/providers/instance/aksmachineinstanceutils.go

```diff
@@ -17,12 +17,274 @@ limitations under the License.
+	if err := validateRetrievedAKSMachineBasicProperties(aksMachine); err != nil {
+		return nil, fmt.Errorf("failed to validate AKS machine instance %q: %w", lo.FromPtr(aksMachine.Name), err)
+	}
+	var zonePtr *string // This one is optional.
+	if len(aksMachine.Zones) < 1 || aksMachine.Zones[0] == nil {
+		log.FromContext(ctx).Info("AKS machine instance is missing zone", "aksMachineName", lo.FromPtr(aksMachine.Name))
+	} else {
+		zonePtr = lo.ToPtr(utils.MakeAKSLabelZoneFromARMZone(aksMachineLocation, lo.FromPtr(aksMachine.Zones[0])))
+	}
+	if aksMachine.Properties.Priority == nil {
+		return nil, fmt.Errorf("AKS machine instance %q is missing priority", lo.FromPtr(aksMachine.Name))
+	}
+
+	return BuildNodeClaimFromAKSMachineTemplate(
+		ctx,
+		aksMachine,
+		offerings.GetInstanceTypeFromVMSize(lo.FromPtr(aksMachine.Properties.Hardware.VMSize), possibleInstanceTypes),
+		GetCapacityTypeFromAKSScaleSetPriority(lo.FromPtr(aksMachine.Properties.Priority)),
+		zonePtr,
+		lo.FromPtr(aksMachine.ID),
+		lo.FromPtr(aksMachine.Properties.ResourceID),
+		IsAKSMachineDeleting(aksMachine),
+		lo.FromPtr(aksMachine.Properties.NodeImageVersion), // Empty: not fatal, no need to check
+	)
+}
+
+// May return apimachinery.NotFoundError if NodePool is not found.
+func FindNodePoolFromAKSMachine(ctx context.Context, aksMachine *armcontainerservice.Machine, kubeClient client.Client) (*karpv1.NodePool, error) {
+	// ASSUMPTION: NodePool name is stored in the all AKS machine tags.
+	nodePoolName, ok := aksMachine.Properties.Tags[launchtemplate.NodePoolTagKey]
+	if ok && *nodePoolName != "" {
+		nodePool := &karpv1.NodePool{}
+		if err := kubeClient.Get(ctx, types.NamespacedName{Name: *nodePoolName}, nodePool); err != nil {
+			return nil, err
+		}
+		return nodePool, nil
+	}
+
+	return nil, errors.NewNotFound(schema.GroupResource{Group: coreapis.Group, Resource: "nodepools"}, "")
+}
+
+// ASSUMPTION: NodeClaim name is in the format of <NodePool name>-<hash suffix>
+// If total length exceeds AKS machine name limit, the exceeded part will be replaced with another deterministic hash.
+// E.g., "thisisalongnodepoolname-a1b2c" --> "thisisalongnoz9y8x7-a1b2c"
+func GetAKSMachineNameFromNodeClaimName(nodeClaimName string) (string, error) {
+	const maxAKSMachineNameLength = 35 // Defined by AKS machine API.
+	const prefixHashLength = 6         // The length of the hashed part replacing the exceeded part of the prefix.
+	// If 6, given alphanumeric hash, there will be a total of 36^6 = 2,176,782,336 combinations.
+
+	if len(nodeClaimName) <= maxAKSMachineNameLength {
+		// Safe to use the whole name
+		return nodeClaimName, nil
+	}
+	splitted := strings.Split(nodeClaimName, "-")
+	// Combine the parts except the last one (NodeClaim hash suffix)
+	prefix := strings.Join(splitted[:len(splitted)-1], "-")
+	suffix := "-" + splitted[len(splitted)-1]
+
```

Comment:
Maybe should guard against name with no hyphen? (I know, not likely ...) Also "split" :-) Actually, would this work:

```suggestion
lastDash := strings.LastIndex(nodeClaimName, "-")
if lastDash == -1 {
    return "", fmt.Errorf("invalid NodeClaim name format: no dash found in %q", nodeClaimName)
}
prefix := nodeClaimName[:lastDash]
suffix := nodeClaimName[lastDash:]
```